lib/url.c

```diff
@@ -1587,19 +1465,229 @@ const struct Curl_handler *Curl_get_scheme_handler(const char *scheme)
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0},
+#if !defined(CURL_DISABLE_SMB) && defined(USE_CURL_NTLM_CORE) && \
+    (SIZEOF_CURL_OFF_T > 4)
+    { &Curl_handler_smb, 121346},
+#else
+    {NULL, 0},
+#endif
+#if defined(USE_SSL) && !defined(CURL_DISABLE_SMB) && \
+    defined(USE_CURL_NTLM_CORE) && (SIZEOF_CURL_OFF_T > 4)
+    { &Curl_handler_smbs, 3883187},
+#else
+    {NULL, 0},
+#endif
+#ifdef USE_LIBRTMP
+    { &Curl_handler_rtmp, 3857936},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+    {NULL, 0}, {NULL, 0},
+#if defined(USE_SSL) && !defined(CURL_DISABLE_POP3)
+    { &Curl_handler_pop3s, 121194195},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0},
+#if defined(USE_SSH)
+    { &Curl_handler_sftp, 3876592},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0},
+#ifndef CURL_DISABLE_DICT
+    { &Curl_handler_dict, 3387604},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+#ifndef CURL_DISABLE_FTP
+    { &Curl_handler_ftp, 108272},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+#if defined(USE_SSH) && !defined(USE_WOLFSSH)
+    { &Curl_handler_scp, 121040},
+#else
+    {NULL, 0},
+#endif
+#if defined(USE_SSL) && !defined(CURL_DISABLE_IMAP)
+    { &Curl_handler_imaps, 113775219},
+#else
+    {NULL, 0},
+#endif
+#ifndef CURL_DISABLE_TELNET
+    { &Curl_handler_telnet, 4001875220},
+#else
+    {NULL, 0},
+#endif
+#ifdef USE_LIBRTMP
+    { &Curl_handler_rtmpe, 123454053},
+#else
+    {NULL, 0},
+#endif
+#ifndef CURL_DISABLE_MQTT
+    { &Curl_handler_mqtt, 3691252},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+#if defined(USE_WEBSOCKETS) && \
+    defined(USE_SSL) && !defined(CURL_DISABLE_HTTP)
+    { &Curl_handler_wss, 125651},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0},
+#if defined(USE_SSL) && !defined(CURL_DISABLE_FTP)
+    { &Curl_handler_ftps, 3464819},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0},
+#ifndef CURL_DISABLE_POP3
+    { &Curl_handler_pop3, 3787315},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0},
+#ifdef USE_LIBRTMP
+    { &Curl_handler_rtmps, 123454067},
+#else
+    {NULL, 0},
+#endif
+#ifdef USE_LIBRTMP
+    { &Curl_handler_rtmpt, 123454068},
+#else
+    {NULL, 0},
+#endif
+    {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {NULL, 0},
+    {NULL, 0}, {NULL, 0}, {NULL, 0},
+  };
+
+  if(len > 7)
+    /* no internal scheme is this long */
+    return NULL;
+
+  while(len) {
+    c <<= 5;
+    c += Curl_raw_tolower(*scheme);
+    scheme++;
+    len--;
+  }
+
+  e = (int)(c % 118);
+  if(i[e].h && (i[e].hash == c))
+    return i[e].h;
+  return NULL;
```

Comment:
sorry, I switched away from the version that stores the hash, because we have functions that will allow "any" characters in the string (not just ascii) so there is still a risk for a hash collision, which a single case compare will avoid